lib/checks/tables/td-headers-attr-evaluate.js

```diff
@@ -3,25 +3,32 @@ import { isVisibleToScreenReaders } from '../../commons/dom';
 
 export default function tdHeadersAttrEvaluate(node) {
```

Comment:
I took a stab at refactoring this function. This was already looping over headers more than it needed to, and with the proposed change it would be twice. Let me know if you like this direction:

```js
import { tokenList } from '../../core/utils';
import { isVisibleToScreenReaders } from '../../commons/dom';
import { getRole } from '../../commons/aria';

// Order determines the priority of reporting
// Only if 0 of higher issues exists will the next be reported
const messageKeys = [
  'cell-header-not-in-table',
  'cell-header-not-th',
  'header-refs-self',
  'empty-hdrs', // incomplete
]
const [notInTable, notTh, selfRef, emptyHdrs] = messageKeys;

export default function tdHeadersAttrEvaluate(node) {
  const cells = [];
  const cellRoleById = {}
  for (let rowIndex = 0; rowIndex < node.rows.length; rowIndex++) {
    const row = node.rows[rowIndex];

    for (let cellIndex = 0; cellIndex < row.cells.length; cellIndex++) {
      const cell = row.cells[cellIndex];
      cells.push(cell);

      // Save header id to set if it's th or td with roles columnheader/rowheader
      const cellId = cell.getAttribute('id');
      if (cellId) {
        cellRoleById[cellId] = getRole(cell);
      }
    }
  }

  const badCells = { [selfRef]: [], [notInTable]: [], [notTh]: [], [emptyHdrs]: [] }
  cells.forEach(cell => {
    if (!cell.hasAttribute('headers') || !isVisibleToScreenReaders(cell)) {
      return;
    }
    const headersAttr = cell.getAttribute('headers').trim();
    if (!headersAttr) {
      badCells[emptyHdrs].push(cell);
      return;
    }

    const cellId = cell.getAttribute('id');
    // Get a list all the values of the headers attribute
    const headers = tokenList(headersAttr);
    headers.forEach(headerId => {
      if (cellId && headerId === cellId) {
        // Header references its own cell
        badCells[selfRef].push(cell);
      } else if (!cellRoleById[headerId]) {
        // Header references a cell that is not in the table
        badCells[notInTable].push(cell);
      } else if (['columnheader', 'rowheader'].includes(cellRoleById[headerId])) {
        // Header references a cell that is not a row or column header
        badCells[notTh].push(cell);
      }
    })
  });

  for (const messageKey of messageKeys) {
    if (badCells[messageKey].length > 0) {
      this.relatedNodes(badCells[messageKey]);
      if (messageKey === emptyHdrs) {
        return undefined;
      }
      this.data({ messageKey });
      return false;
    }
  }
  return true;
}
```

Reply:
Looks better than mine :sweat_smile: I will  take your impl. The only thing in mind i have is better to show all errors if there may so that user will not go back and forward and fix errors one by one, not sure if it is possible?

Reply:
Unfortunately there is no good way for a check to report multiple issues. You're right that would be better. Axe-core's architecture doesn't really allow that.